### PR TITLE
fix(popup): always use current window when position == "cursor"

### PR DESCRIPTION
**Pull request by pysan3**
Fixes https://github.com/folke/noice.nvim/issues/636.

There are many ways to fix this issue, but I think my approach is the cleanest.

Other options:

- Overwrite `internal.position.win` to nil just like `win_config.win` here. (`win_config.win` will be nil, but `internal.position.win` is still unchanged and remembers the old value.)
https://github.com/MunifTanjim/nui.nvim/blob/c3c7fd618dcb5a89e443a2e1033e7d11fdb0596b/lua/nui/layout/utils.lua#L145
- Pass `win_config` instead of `internal.position` to `mod.get_container_info` here.
https://github.com/MunifTanjim/nui.nvim/blob/c3c7fd618dcb5a89e443a2e1033e7d11fdb0596b/lua/nui/layout/utils.lua#L156

Basically we need to pass a table with `win = nil` to `get_window_size()` that is called inside `get_container_info`, or deal with invalid `table.win` and ignore that value inside that function.

> [!NOTE]
> It might be better to `relative ~= "win" and 0 or position.win`?
>
> Basically the values are `relative = "editor"|"win"|"cursor"`, and "editor" case is dealt with above, so they mean the exact same thing for now, but may not be in the future (when new position value is added).

### CODE REVIEW

**Comment by pysan3** [comment]
@MunifTanjim Any updates?

**Comment by MunifTanjim** [comment]
What would happen if I open a popup with relative=cursor, then move the cursor to a different window and call `popup:update_layout`? 🤔

**Comment by pysan3** [comment]
Oh, I thought `position = "cursor"` should follow the cursor. So when `update_layout`, the popup position is remapped to the position / window of which the cursor is positioned right now.

---

This is how to reproduce the problem I was referring to.

And this PR fixes it by making popup always open from the current focused window.

1. new popup (relative = "cursor", position = { row = 1, col = 1 })
2. close it
3. `:vs`
4. `:e different file`
5. reopen the popup
6. `<C-w>w` (go back to the previous window)
7. `:q` (close this window)
8. (now there's only the new window opened by `:vs`)
9. reopen the popup
10. `ERROR ...o/.local/share/nvim/lazy/nui.nvim/lua/nui/utils/init.lua:29: Invalid window id: 1000`

---

When `relative = "win"` I agree that the popup should stay on the window where it was initially spawned from even after `:update_layout()` (and do not follow the `nvim_get_current_win()`).

**Comment by pysan3** [comment]
Any updates @MunifTanjim ?

File: lua/nui/layout/utils.lua [hunk 1]
```diff
@@ -89,7 +89,7 @@ function mod.get_container_info(position)
 
   return {
     relative = position.bufpos and "buf" or relative,
-    size = utils.get_window_size(position.win),
+    size = utils.get_window_size(relative == "cursor" and 0 or position.win),
```
Comment (MunifTanjim) [inline lua/nui/layout/utils.lua:92]: This will just fix the size. But reference to that invalid window id will remain internally.

Can we try this instead?

```diff
diff --git a/lua/nui/layout/utils.lua b/lua/nui/layout/utils.lua
index 6f251b5..66efcc3 100644
--- a/lua/nui/layout/utils.lua
+++ b/lua/nui/layout/utils.lua
@@ -137,7 +137,9 @@ function mod.update_layout_config(component_internal, config)
   if options.relative then
     internal.layout.relative = options.relative
 
-    local fallback_winid = internal.position and internal.position.win or vim.api.nvim_get_current_win()
+    local fallback_winid = internal.position and internal.position.win
+      or internal.layout.relative.type == "cursor" and 0
+      or vim.api.nvim_get_current_win()
     internal.position =
       vim.tbl_extend("force", internal.position or {}, mod.parse_relative(internal.layout.relative, fallback_winid))
 
```

Comment (pysan3) [reply lua/nui/layout/utils.lua:92]: Fair enough. I'll use that instead.

What do you think about this?

> [!NOTE]
> It might be better to `relative ~= "win" and 0 or position.win`?
>
> Basically the values are `relative = "editor"|"win"|"cursor"`, and "editor" case is dealt with above, so they mean the exact same thing for now, but may not be in the future (when new position value is added).

Comment (MunifTanjim) [reply lua/nui/layout/utils.lua:92]: `nui.nvim` also accepts this: https://github.com/MunifTanjim/nui.nvim/blob/fc59553b5a8a1c13b8aa25ae62b6a47ec2b1882c/lua/nui/popup/README.md#L229-L240

Which internally translates to

```
win = current_window
bufpos = position_relative_to_buffer_on_current_window
```

The `bufpos` is only valid for a specific buffer, so `win` needs to be fixed to a window that contains that buffer.

Comment (pysan3) [reply lua/nui/layout/utils.lua:92]: Oh so it is `relative = "editor"|"win"|"cursor"|"buf"`?

Then I'll use `relative == "cursor" and 0`.

Comment (pysan3) [reply lua/nui/layout/utils.lua]: @MunifTanjim Fixed!

**Review by MunifTanjim — APPROVED** [review]
